src/main/java/org/tarantool/TarantoolClientImpl.java

```diff
@@ -461,33 +505,142 @@ protected void writeThread() {
+            request.setCompletedSchemaId(schemaId);
+            if (request.isSync()) {
+                completeSync(request);
+            } else if (request.getCode() == Code.EXECUTE) {
+                completeSql(request, packet);
             } else {
-                Object error = packet.getBody().get(Key.ERROR.getId());
-                fail(future, serverError(code, error));
+                ((TarantoolOp) result).complete(packet.getData());
+            }
+        } else if (code == ProtoConstants.ERR_WRONG_SCHEMA_VERSION) {
+            if (schemaId > schemaMeta.getSchemaVersion()) {
+                delayedOpsQueue.add(request);
+            } else {
+                registerOperation(request, schemaMeta.getSchemaVersion());
+            }
+        } else {
+            Object error = packet.getError();
+            fail(request, serverError(code, error));
+        }
+
+        if (request.getStartedSchemaId() == 0) {
+            return;
+        }
+        // it's possible to receive bigger version than current
+        // i.e. after DDL operation or wrong schema version response
+        if (schemaId > schemaMeta.getSchemaVersion()) {
+            updateSchema();
+        }
+    }
+
+    /**
+     * When a sync is successful it's required
+     * to re-evaluate dependent request and complete
+     * exceptionally with a proper error.
+     *
+     * @param request sync request
+     */
+    private void completeSync(TarantoolRequest request) {
+        TarantoolRequest source = request.getSyncDependent();
+        if (delayedOpsQueue.remove(source)) {
+            try {
+                source.getArguments();
+            } catch (Exception e) {
+                fail(source, e);
+            }
+        }
+    }
+
+    /**
+     * Optimistic means the client issues ping request
+     * to check whether it keeps an actual schema version or not.
+     * If the schema is outdated, client uses refreshes its schema
+     * cache.
+     * <p>
+     * It's suitable when we don't refresh all the cache directly because
+     * of the schema can be large enough to be loaded fast.
+     *
+     * @see #updateSchema()
+     */
+    private void optimisticSchemaUpdate(TarantoolRequest source) {
+        TarantoolRequest pingRequest = new TarantoolRequest(Code.PING);
+        pingRequest.beginSync(syncId.incrementAndGet(), operationTimeout, source);
+        registerOperation(pingRequest, schemaMeta.getSchemaVersion());
+    }
+
+    private void updateSchema() {
+        performSchemaAction(() -> {
+            if (state.acquire(StateHelper.SCHEMA_UPDATING)) {
+                workExecutor.execute(createUpdateSchemaTask());
+            }
+        });
+    }
+
+    private Runnable createUpdateSchemaTask() {
+        return () -> {
+            try {
+                schemaMeta.refresh();
+            } catch (Exception cause) {
+                workExecutor.schedule(createUpdateSchemaTask(), 300L, TimeUnit.MILLISECONDS);
```

Comment:
> what a behaviour we'll see if, say, a connectivity problem will appear during a schema update?

Because schema updating is a scheduler task and it uses same client then any connectivity issue will lead the same reconnection attempts. `However, the task will be interrupted by an exception from the client (completes). As a result, `workExecutor` doesn't suffer from this kind of problem and can be continued to use after the reconnection.

Reply:
> However, the task will be interrupted by an exception from the client (completes).

But the exception is catched and the new schema update task will be scheduled: this is what I called 'loop' above.

If we'll unable to connect and the client will step into the died state, will the 'loop' stop?

Reply:
well, the endless loop is impossible here. If a client fails to (re-)connect to all the available instances and a channel provider returns not a transient error the client will close itself internally which causes the final CLOSE state set and the executor to be shutdown.

https://github.com/tarantool/tarantool-java/blob/dd1f29fd340b2150213712fcfe3b67988ed787db/src/main/java/org/tarantool/TarantoolClientImpl.java#L184-L191

Reply:
Thanks, I see now: a reader or writer thread will call `die()`, then the thread will signal for reconnection to a connector thread, which will call `reconnect()`, which will call `close()` if we're unable to connect at all. So the executor will be closed, you are right.

However now I have another question: we'll fail all futures (both in-fly and delayed ones) before even attempt to reconnect, in `die()` called from reader / writer thread. I would expect that the client will resend them after succussful reconnection. Am I right that now it is not so? If so, is it non-intentional behaviour? And, if so, whether it is the regression?

(Of course, this is not something that was introduced here; I just found this while looking into your answer.)

Reply:
It depends on how we take a look at this. It's not a regression 'cause it never worked in this way. 
Here a simple client does not resend requests and makes dead already registered ones. On the other hand, a cluster client does it in a similar but limited way: resend only requests that were issued during the reconnection step.

https://github.com/tarantool/tarantool-java/blob/dd1f29fd340b2150213712fcfe3b67988ed787db/src/main/java/org/tarantool/TarantoolClusterClient.java#L126-L136

and

https://github.com/tarantool/tarantool-java/blob/dd1f29fd340b2150213712fcfe3b67988ed787db/src/main/java/org/tarantool/TarantoolClusterClient.java#L166-L184

Looks it was implemented intentionally in a simple client (relying on the cluster client) but was implemented partly in clustered one.

Reply:
Interesting. Thanks!

Anyway, it is not part of the scope of this PR. Closed the conversation.